components/percona_telemetry/data_provider.cc

```diff
@@ -120,25 +134,29 @@ bool DataProvider::do_query(const std::string &query, QueryResult *result,
   }
   result->clear();
 
+  /* command_factory_service_.init() allocates memory for mysql_h
+    We need to call close() always.
+    Even if init() fails, becaues it doesn't allocate anything, calling close()
+    is safe, because internally it checks if provided pointer is valid
+  */
+  std::shared_ptr<MYSQL_H> mysql_h_close_guard(
+      &mysql_h, [&srv = command_factory_service_](MYSQL_H *ptr) {
+        srv.close(*ptr);
+      });
+
   mysql_service_status_t sstatus = command_factory_service_.init(&mysql_h);
```

Comment:
Another improvement I discovered recently, which may help simplify things a bit.
Oracle recently introduced a new `MYSQL_COMMAND_LOCAL_THD_HANDLE` option for `mysql_command_service`.
Basically, then you set this option with `nullptr` value, it will create an internal THD object, initialize its session vars, calls my_thread_init() / my_thread_end(), sets `current_thd`,  etc. This THD object will be destroyed when you call `mysql_command_service::close()`.
https://github.com/percona-ysorokin/percona-server/commit/c9a98d3ceacf37449d14b229a9e4e9e2f94138e0#diff-3706ffbbf71661ebd0414b16d27fd6a1cc8f6960488d43d5ad51d9b15fe9ac00R51